### PR TITLE
fix(transition): animate all toolbars in iOS page transitions

### PR DESCRIPTION
**Pull request by zakton5**
#### Short description of what this resolves:
Only the first `ion-toolbar` within an `ion-header` animates when navigating forward or back.

#### Changes proposed in this pull request:
- Add all toolbars within `ion-header` to the transition instead of just one

**Ionic Version**: 4.0.0@rc.3

**Fixes**: #16262

### CODE REVIEW

**Comment by DavidWiesner** [comment]
This is the wrong way to fix this issue

File: core/src/utils/transition/ios.transition.ts [hunk 1]
```diff
@@ -41,10 +41,11 @@ export function iosTransitionAnimation(AnimationC: Animation, navEl: HTMLElement
   // setting up enter view
   const contentEl = enteringEl.querySelector(':scope > ion-content');
   const headerEls = enteringEl.querySelectorAll(':scope > ion-header > *:not(ion-toolbar), :scope > ion-footer > *');
```
[the remark this answers is not in the document]
Comment (zakton5) [reply core/src/utils/transition/ios.transition.ts]: That could work @DavidWiesner. However, lines 68-120 modify the animation in a very specific way. Here is what is looks like with your suggestion below. Notice the title in the first header lags behind the contents of the second header.

toolbar_transition

If that is desired behavior I will change it.

**Comment by jayordway** [comment]
I am excited to see this issue fixed! Hopefully it is merged soon!

**Comment by DavidWiesner** [comment]
There is another issue in the animation. This was also an issue before your pull request. What if there is an other element e.g.: an `ion-searchbar` in one of the toolbars.
E.g.:
```
<ion-header>
  <ion-toolbar color="secondary">
    <ion-buttons slot="start">
      <ion-back-button></ion-back-button>
    </ion-buttons>
    <ion-title>SecondPage</ion-title>
  </ion-toolbar>
  <ion-toolbar>
    <ion-searchbar></ion-searchbar>
  </ion-toolbar>
</ion-header>
```
Maybe you can fix this issue also in your pull request.

**Comment by elmartino** [comment]
+1

**Comment by zakton5** [comment]
@DavidWiesner Good catch! I'll play around with the selectors and see what I can find. 
I know adding each element to the animation isn't the most efficient, but I'm just extending the logic that was already there. I think the extra overhead to the animation is important to maintain the "feel" of iOS.

**Comment by zakton5** [comment]
@DavidWiesner Okay, I think I've got it now. It's still a bit inefficient in terms of adding a bunch of stuff to the animation, but it does the job. And now custom elements/components shouldn't be a problem either. It should handle `ion-searchbar` as well. Here is what it looks like with the most recent updates:

toolbar_transition_2

**Comment by jayordway** [comment]
Will it handle segment buttons and ion-segment being in there as well?

**Comment by zakton5** [comment]
@jayordway Yes i just tried it. It handles that as well. Should handle anything as long as it is placed within a toolbar.

**Comment by jayordway** [comment]
When is this getting merged?

**Comment by denisroldan** [comment]
Still waiting for this 🤞🙄

**Comment by dansiemens** [comment]
Also waiting for this.

**Comment by jayordway** [comment]
So I guess this one missed the 4.1.x release....maybe one of these days!

**Comment by zSkalin** [comment]
How is this not a higher priority? Are people releasing apps with a buggy toolbar?

Ionic 4 is unusable for me until this is fixed, it looks horrible

File: core/src/utils/transition/ios.transition.ts [hunk 2]
```diff
@@ -140,16 +149,20 @@ export function iosTransitionAnimation(AnimationC: Animation, navEl: HTMLElement
...
       const leavingTitle = new AnimationC();
       leavingTitle.addElement(leavingToolBarEle.querySelector('ion-title'));
 
-      const leavingToolBarItems = new AnimationC();
-      leavingToolBarItems.addElement(leavingToolBarEle.querySelectorAll('ion-buttons,[menuToggle]'));
+      const leavingToolBarButtons = new AnimationC();
+      leavingToolBarButtons.addElement(leavingToolBarEle.querySelectorAll('ion-buttons,[menuToggle]'));
+
+      const leavingToolbarItems = new AnimationC();
+      leavingToolbarItems.addElement(leavingToolBarEle.querySelectorAll(':scope > *:not(ion-title):not(ion-buttons):not([menuToggle]'));
```
Comment (liamdebeasi) [inline core/src/utils/transition/ios.transition.ts]: What does this line do?

**Comment by liamdebeasi** [comment]
@zakton5 I pushed an update that has a few changes:

- Fixed a bug for apps running in Safari/WebKit that would cause animations to not run
- Updated the test in `basic` to be a little more comprehensive
- Cleaned up the code a bit

Let me know if you have any concerns! I need to do some performance testing but I will be doing that shortly.

**Comment by liamdebeasi** [comment]
Just as an update: I am noticing some performance issues on older devices when animating more than 1 toolbar. I am currently discussing with other Ionic team members on the best way to address this.

Feel free to jump into the code and see if you can make any optimizations! 🚀

**Comment by pegler** [comment]
Exciting! Would it be possible to get a dev release when this is merged?

**Comment by brandyscarney** [comment]
Thanks everyone who contributed to this! ❤️ We will definitely get a dev release out once merged.

**Comment by liamdebeasi** [comment]
Thank you everyone! 🎉

**Comment by liamdebeasi** [comment]
Hi everyone,

I have pushed a nightly build with these changes (`npm i @ionic/core@dev`)

Thanks!

**Comment by jayordway** [comment]
@liamdebeasi So I have tested with 4.1.2-dev.201903141948.7d01207 and it seems as if this was fixed just for backing out of the view with the "sticky" or static toolbar but not transitioning into it.

**Comment by liamdebeasi** [comment]
Hi @jayordway,

Can you provide the code you used to test this? I can look into it.

Thanks!

**Comment by jayordway** [comment]
@liamdebeasi Due to the nature of the project I cannot make it public, but here is a snippet to show the toolbar section that has the toolbar which appears immediately before the transition/slide even starts:

[<ion-header>
  <ion-toolbar>
    <ion-buttons slot="start">
      <ion-back-button ></ion-back-button>
    </ion-buttons>
    <ion-title>Ticket details</ion-title>
  </ion-toolbar>
  <ng-container *ngIf="(ticket$ | async) as ticket">
    <ion-toolbar  *ngIf="ticket.enableRequestUpdateTicket || ticket.enableRequestCloseTicket
                               || ticket.enableContestTicket || ticket.enableEscalateTicket" class="details-toolbar">
      <ion-buttons class="ticket-action-buttons">
        <ion-button *ngIf="ticket.enableRequestUpdateTicket " [shape]="'round'" [fill]="'outline'"  (click)="presentModal(actionEnum.UPDATE)">Update</ion-button>
        <ion-button *ngIf="ticket.enableRequestCloseTicket"   [shape]="'round'" [fill]="'outline'"  (click)="presentModal(actionEnum.CLOSE)">Req. close</ion-button>
        <ion-button *ngIf="ticket.enableContestTicket"        [shape]="'round'" [fill]="'outline'"  (click)="presentModal(actionEnum.CONTEST)">Contest</ion-button>
        <ion-button *ngIf="ticket.enableEscalateTicket"       [shape]="'round'" [fill]="'outline'"  (click)="presentModal(actionEnum.ESCALATE)">Escalate</ion-button>
      </ion-buttons>
    </ion-toolbar>
  </ng-container>
</ion-header>](url)

**Comment by liamdebeasi** [comment]
Hi there,

When `ion-button` is inside `ion-buttons`, the `ion-button` components will not animate/move (they will only fade in/out). This behavior is consistent with what is found in many iOS apps.

If you feel you have found an bug, please create an issue [here](https://github.com/ionic-team/ionic/issues/new?assignees=&labels=&template=bug_report.md&title=). Thanks!

**Comment by Horst1102** [comment]
Hi,

i am facing problems with the page transition when coming from a page that has multiple toolbars with segments to a page that also has mutliple toolbars (but one more than the page before).

Head of Page 1 (a tabbed page):
```
<ion-header>
    <ion-toolbar>
        <ion-buttons slot="start">
            <ion-menu-button></ion-menu-button>
        </ion-buttons>
        <ion-buttons slot="end">
            <ion-button (click)="updateContent()">
                <ion-icon slot="icon-only" name="refresh"></ion-icon>
            </ion-button>
        </ion-buttons>
        <ion-title>{{ 'REPORTER' | translate }}</ion-title>
    </ion-toolbar>
    <ion-toolbar no-border-top>
        <ion-segment [(ngModel)]="reporterFilter">
            <ion-segment-button value="all">
                {{ 'ALL' | translate }}
            </ion-segment-button>
            <ion-segment-button value="unconfirmed">
                {{ 'UNSEEN' | translate }}
            </ion-segment-button>
        </ion-segment>
    </ion-toolbar>
</ion-header>
```

Head of Page 2 (opened from the tab page):
```
<ion-header>
    <ion-toolbar>
        <ion-buttons slot="start">
            <ion-back-button text=""></ion-back-button>
        </ion-buttons>
        <ion-buttons slot="end">
            <ion-button
                    (click)='navigate(reporterDetailResponse.street + ", " + reporterDetailResponse.postcode + ", " + reporterDetailResponse.city)'>
                <ion-icon slot="icon-only" name="navigate"></ion-icon>
            </ion-button>
            <ion-button (click)="presentFilter()">
                <ion-icon slot="icon-only" name="options"></ion-icon>
            </ion-button>
        </ion-buttons>
        <ion-title>{{'SINGLEREPORTER' | translate}}</ion-title>
    </ion-toolbar>

    <ion-toolbar>
        <ion-segment [(ngModel)]="segmentAnlagenDetail">
![ezgif-3-9cf88866f761](https://user-images.githubusercontent.com/49471647/55857796-5dd0b580-5b6e-11e9-95a1-fbf9440dfc20.gif)


            <ion-segment-button value="alarms">
                {{ 'ALARMS' | translate }}
            </ion-segment-button>
            <ion-segment-button value="comments">
                {{ 'COMMENTS' | translate }}
            </ion-segment-button>
        </ion-segment>
    </ion-toolbar>

    <ion-toolbar no-padding>
        <div class="filter-item filter-caption">{{'FILTER'| translate}}:</div>
        <div *ngFor="let level of getExcludedLevel()">
            <ion-badge class="filter-item" (click)="removeLevel(level)">{{level.name | translate}}</ion-badge>
        </div>
        <ion-badge class="filter-item" *ngIf="fromDate" (click)="removeDate()">
            <ion-icon name="calendar"></ion-icon>
        </ion-badge>
        <ion-badge class="filter-item" *ngIf="userOnly" (click)="removeUserOnly()">
            <ion-icon name="person"></ion-icon>
        </ion-badge>
    </ion-toolbar>

</ion-header>
```


So my problem is the following: 
- when changing from page 1 to page 2 the first list item of the content from page 1 is shown in the third toolbar of page 2 while animating (fading in/fasding out).
- when changing back to page 1 from page 2 the third toolbar ist shown in the first list item ...

I have added a gif for you.
ezgif-3-9cf88866f761


How can i prevent this?

**Comment by jayordway** [comment]
@Horst1102 Didn't Liam state in the previous comment that ion-buttons will not move with the transition?

**Comment by Horst1102** [comment]
> @Horst1102 Didn't Liam state in the previous comment that ion-buttons will not move with the transition?

Yes, i have read that. But as far as i uderstand i do not use ion-buttons in the third toolbar. 

The element that overlaps is no button:  `<div class="filter-item filter-caption">{{'FILTER'| translate}}:</div>`

**Comment by liamdebeasi** [comment]
Hi there,

If you feel you have found a bug, please open a new [Bug Report](https://github.com/ionic-team/ionic/issues/new?assignees=&labels=&template=bug_report.md&title=bug%3A+).

**Pull request merged**